x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx

```diff
@@ -474,42 +509,56 @@ export class PrivilegeSpaceForm extends Component<Props, State> {
   };
 
   private onFeaturePrivilegesChange = (featureId: string, privileges: string[]) => {
-    const role = copyRole(this.state.role);
-    const form = role.kibana[this.state.privilegeIndex];
-
-    if (privileges.length === 0) {
-      delete form.feature[featureId];
-    } else {
-      form.feature[featureId] = [...privileges];
-    }
-
-    this.setState({
-      role,
-      privilegeCalculator: new PrivilegeFormCalculator(this.props.kibanaPrivileges, role),
-    });
+    this.setRole(privileges, featureId);
   };
 
   private onChangeAllFeaturePrivileges = (privileges: string[]) => {
+    this.setRole(privileges);
+  };
+
+  private setRole(privileges: string[], featureId?: string) {
     const role = copyRole(this.state.role);
     const entry = role.kibana[this.state.privilegeIndex];
 
     if (privileges.length === 0) {
-      entry.feature = {};
+      if (featureId) {
+        delete entry.feature[featureId];
+      } else {
+        entry.feature = {};
+      }
     } else {
-      this.props.kibanaPrivileges.getSecuredFeatures().forEach((feature) => {
+      let securedFeaturesToSet = this.props.kibanaPrivileges.getSecuredFeatures();
+      if (featureId) {
+        securedFeaturesToSet = [securedFeaturesToSet.find((sf) => sf.id === featureId)!];
+      }
+      securedFeaturesToSet.forEach((feature) => {
         const nextFeaturePrivilege = feature
-          .getPrimaryFeaturePrivileges()
-          .find((pfp) => privileges.includes(pfp.id));
+          .getPrimaryFeaturePrivileges({ includeMinimalFeaturePrivileges: true })
+          .find((pfp) => {
+            if (
+              pfp?.disabled ||
+              (pfp?.requireAllSpaces && !this.state.selectedSpaceIds.includes(ALL_SPACES_ID))
+            ) {
+              return false;
+            }
+            return Array.isArray(privileges) && privileges.includes(pfp.id);
+          });
         if (nextFeaturePrivilege) {
```

Comment:
This still suffers from the problem we identified yesterday, where we aren't unsetting the `All` privilege for Fleet when bulk-changing from `All` to `Read`:

1) Configure Fleet to `requireAllSpaces` for the `all` privilege
2) Disable Fleet's `read` privilege
3) Go to role management, start to create a new role
4) Add Kibana privilege for All Spaces
5) Bulk change to `All`
6) Bulk change to `Read`

Expect: Fleet changes from `All` to `None`
Actual: Fleet stays at `All`

Reply:
Yes, I messed up here, I tried to do something cleaner and I got it wrong